### PR TITLE
dnstap support for kresd

### PR DESCRIPTION
**Pull request by vendemiat**

### CODE REVIEW

**Comment by edmonds** [comment]
Hi,

Thanks for doing this! (Original author of fstrm here.)

Would it make sense to add a `--with-library-only` flag to fstrm's configure script so that you don't have to drag in libevent, too? libevent is only used in the `fstrm_capture` utility, not in the core library.

Also, I'm not that familiar with the kresd build system, but it seems like depending on specific versions of `libprotobuf-c` and `libfstrm` would be fragile:

```
$(eval $(call find_lib,libprotobuf-c,1.2.1))
$(eval $(call find_lib,libfstrm,0.2))
```

These libraries have stable ABIs, so it should be possible to depend on their SOVERSION only, e.g. `libprotobuf-c.so.1` and `libfstrm.so.0`.

BTW, the latest release of `fstrm` is `0.3.0`.

**Comment by vendemiat** [comment]
Thanks Edmonds. Will make the suggested changes and also test against fstrm 0.3.0.

**Comment by vavrusa** [comment]
The `find_lib,libprotobuf-c,1.2.1` has `pkg-config --atleast-version` semantics [1], so it should be okay .. until there is a breaking change in the upstream and we would need to introduce `--max-version` semantics too.

[1]: https://github.com/CZ-NIC/knot-resolver/blob/master/platform.mk#L145

File: lib/resolve.h [hunk 1]
```diff
@@ -123,6 +123,7 @@ struct kr_request {
         const struct sockaddr *dst_addr;
         const knot_pkt_t *packet;
         const knot_rrset_t *opt;
+        bool tcp; /* true if the query is on tcp */
```
Comment (vavrusa) [inline lib/resolve.h]: ^ This will likely need Lua headers regenerated @vcunat

File: modules/dnstap/dnstap.c [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+		return NULL;
+	}
+	sbuf.must_free_data = true;
+
+	*sz = dnstap__dnstap__pack_to_buffer(d, (ProtobufCBuffer *) &sbuf);
+	*buf = sbuf.data;
+	return *buf;
+}
+
+/* set_address fills in address deatail in dnstap_message
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: ^ s/deatail/detail/

File: modules/dnstap/dnstap.c [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+		addr->len = sizeof(sai->sin_addr);
+		addr->data = (uint8_t *)&sai->sin_addr.s_addr;
+		*port = ntohs(sai->sin_port);
+	} else if (sockaddr->sa_family == AF_INET6) {
+		const struct sockaddr_in6 *sai6;
+		sai6 = (const struct sockaddr_in6 *)sockaddr;
+		addr->len = sizeof(sai6->sin6_addr);
+		addr->data = (uint8_t *)&sai6->sin6_addr.s6_addr;
+		*port = ntohs(sai6->sin6_port);
+	}
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: There are convenience functions for this, please reuse:
* https://github.com/CZ-NIC/knot-resolver/blob/master/lib/utils.c#L245-L270
* https://github.com/CZ-NIC/knot-resolver/blob/master/lib/utils.h#L129-L137
(and add similar one for port)

File: modules/dnstap/dnstap.c [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+		const knot_pkt_t *rpkt = req->answer;
+		m.response_message.len = rpkt->size;
+		m.response_message.data = (uint8_t *)rpkt->wire;
+		m.has_response_message = true;
+	}
+
+	/* set query time to the timestamp of the first kr_query
+	 * set response time to now
+	 */
+	if (rplan->resolved.len > 0 ) {
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: Trailing whitespace after `0`

File: modules/dnstap/dnstap.c [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+		m.query_time_sec = first->timestamp.tv_sec;
+		m.has_query_time_sec = true;
+		m.query_time_nsec = first->timestamp.tv_sec * 1000;
+		m.has_query_time_nsec = true;
+	}
+
+	/* Response time */
+	m.response_time_sec = now.tv_sec;
+	m.has_response_time_sec = true;
+	m.response_time_nsec = now.tv_sec * 1000;
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: Isn't `response_time_nsec` the remaining fraction of a second?
Either way, you should use `now.tv_nsec`.

File: modules/dnstap/dnstap.c [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+
+	/* set query time to the timestamp of the first kr_query
+	 * set response time to now
+	 */
+	if (rplan->resolved.len > 0 ) {
+		struct kr_query *first = rplan->resolved.at[0];
+
+		m.query_time_sec = first->timestamp.tv_sec;
+		m.has_query_time_sec = true;
+		m.query_time_nsec = first->timestamp.tv_sec * 1000;
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: Use `.tv_nsec`

File: modules/dnstap/dnstap.c [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+	/* Response time */
+	m.response_time_sec = now.tv_sec;
+	m.has_response_time_sec = true;
+	m.response_time_nsec = now.tv_sec * 1000;
+	m.has_response_time_nsec = true;
+
+	/* Qname */
+	m.query_zone.data = (uint8_t *)knot_pkt_qname(req->answer);
+	m.query_zone.len = req->answer->qname_size;
+	m.has_query_zone = true;
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: `query_zone` should be recovered from current zone cut, like here: https://github.com/CZ-NIC/knot-resolver/blob/master/lib/resolve.c#L1021

File: modules/dnstap/dnstap.c [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+	return kr_ok();
+}
+
+/* dnstap_unix_writer returns a unix fstream writer
+ * https://gitlab.labs.nic.cz/labs/knot/blob/master/src/knot/modules/dnstap.c#L159
+ */
+static struct fstrm_writer* dnstap_unix_writer(const char *path) {
+	struct fstrm_unix_writer_options *opt = NULL;
+	struct fstrm_writer_options *wopt = NULL;
+	struct fstrm_writer *writer = NULL;
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: It looks okay, but can you use cleanup handlers for opt + wopt instead?
See https://github.com/CZ-NIC/knot-resolver/blob/master/contrib/cleanup.h#L25-L28 and https://github.com/CZ-NIC/knot-resolver/blob/30eb8bd5c58e3230439e0e85aa268e9104d7002e/lib/cache.c#L234-L237

File: modules/dnstap/dnstap.c [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+}
+
+/* find_string 
+ * create a new string from json
+ * *var is set to pointer of new string
+ * node must of type JSON_STRING
+ * new string can be at most len bytes
+ */
+static int find_string(const JsonNode *node, char **val, size_t len) {
+	if (!node || !node->key ) {
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: Trailing whitespace after `->key` (in all json convenience functions below as well).

File: modules/dnstap/dnstap.c [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+	assert(node->tag == JSON_BOOL);
+	return node->bool_;
+}
+
+/* parse config */
+KR_EXPORT
+int dnstap_config(struct kr_module *module, const char *conf) {
+	struct dnstap_data *data = module->data;
+	char *sock_path = NULL;
+	struct fstrm_writer *writer;
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: Declare `writer` with initialisation, or even better declare on first use.

File: modules/dnstap/dnstap.c [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+	}
+	assert(node->tag == JSON_BOOL);
+	return node->bool_;
+}
+
+/* parse config */
+KR_EXPORT
+int dnstap_config(struct kr_module *module, const char *conf) {
+	struct dnstap_data *data = module->data;
+	char *sock_path = NULL;
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: Use `auto_free` on `sock_path` and remove all manual frees.

File: modules/dnstap/dnstap.c [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+		DEBUG_MSG("can't init fstrm_iothr\n");
+		fstrm_writer_destroy(&writer);
+		return kr_error(ENOMEM);
+	}
+
+	/* Get fstrm thread handle
+	 * We only have one input queue, hence idx=0
+	 */
+	data->ioq = fstrm_iothr_get_input_queue_idx(data->iothread, 0);
+	if (!data->ioq) {
```
Comment (vavrusa) [inline modules/dnstap/dnstap.c]: `data->iothread` should be closed here I think.

File: lib/resolve.h (hunk 1 above)
Comment (vcunat) [inline lib/resolve.h]: Yes, that's a trivial step now, doable anytime.

**Comment by pspacek** [comment]
@vavrusa Are you okay with the new version?

File: Makefile [hunk 13]
```diff
@@ -69,6 +71,12 @@ endif
 endif
 endif
 
+# check for fstrm and protobuf for dnstap
+ifeq ($(HAS_libfstrm)|$(HAS_libprotobuf-c),yes|yes)
+BUILD_CFLAGS += -DENABLE_DNSTAP
```
Comment (vcunat) [inline Makefile]: This `#define` seems unused.  Am I missing something?

**Comment by vcunat** [comment]
Is there a reason to:
 - include those generated `dnstap.pb-c.*` instead of generating them during make?
 - link those into the contrib target instead of the module itself?

**Comment by vcunat** [comment]
I implemented those ideas atop, on https://gitlab.labs.nic.cz/knot/resolver/merge_requests/213.  I'm inclined to trust you on most of the contents of the module itself.  The interfacing with kresd etc. seems fine to me.

**Pull request closed without merging**